contracts/Strategy.sol

```diff
@@ -53,80 +138,204 @@ contract Strategy is BaseStrategy {
-
         uint256 totalAssets = want.balanceOf(address(this));
         if (_amountNeeded > totalAssets) {
-            _liquidatedAmount = totalAssets;
-            _loss = _amountNeeded.sub(totalAssets);
+            uint256 amountToFree = _amountNeeded.sub(totalAssets);
+
+            (, uint256 lastAccrued, ) = kashiPair.accrueInfo();
+            if (block.timestamp > lastAccrued) {
+                // We need to call accrue to accurately calculate totalAssets
+                kashiPair.accrue();
+            }
+
+            uint256 deposited = estimatedTotalAssets().sub(totalAssets);
+
+            if (deposited < amountToFree) {
+                amountToFree = deposited;
+            }
+            if (deposited > 0) {
+                uint256 sharesToFree =
+                    bentoBox.toShare(
+                        BIERC20(address(want)),
+                        amountToFree,
+                        true
+                    );
+                uint256 sharesToFreeFromKashi =
+                    sharesToFree.sub(
+                        bentoBox.balanceOf(
+                            BIERC20(address(want)),
+                            address(this)
+                        )
+                    );
+
+                if (sharesToFreeFromKashi > 0) {
+                    kashiPair.removeAsset(
+                        address(this),
+                        _toKashiFraction(sharesToFreeFromKashi, true)
+                    );
+                }
+
+                bentoBox.withdraw(
+                    BIERC20(address(want)),
+                    address(this),
+                    address(this),
+                    0,
+                    bentoBox.balanceOf(BIERC20(address(want)), address(this))
+                );
+            }
+
+            _liquidatedAmount = want.balanceOf(address(this));
         } else {
             _liquidatedAmount = _amountNeeded;
         }
     }
 
     function liquidateAllPositions() internal override returns (uint256) {
-        // TODO: Liquidate all positions and return the amount freed.
+        liquidatePosition(type(uint256).max);
         return want.balanceOf(address(this));
     }
 
-    // NOTE: Can override `tendTrigger` and `harvestTrigger` if necessary
-
     function prepareMigration(address _newStrategy) internal override {
-        // TODO: Transfer any non-`want` tokens to the new strategy
-        // NOTE: `migrate` will automatically forward all `want` in this strategy to the new one
+        liquidateAllPositions();
+    }
+
+    function setKashiPair(address _newKashiPair) external onlyAuthorized {
+        require(
+            IKashiPair(_newKashiPair).bentoBox() == bentoBox,
+            "BentoBox does not match"
+        );
+        require(
+            IKashiPair(_newKashiPair).asset() == BIERC20(address(want)),
+            "KashiPair asset does not match want"
+        );
+
+        liquidateAllPositions();
```

Comment:
weird to see a liquidate here. Why would we set a new kashi pair, instead of cloning a new strat?

Reply:
Good question, let me refactor.

Reply:
fixed